### ISSUE TITLE
discuss installation/execution/deployment model

### ISSUE THREAD

**Issue by gernotstarke** [docToolchain/docToolchain#278]
It has been discussed aside various other issues:

What would be a good installation or deployment model for docToolchain?

[] In [issue #259](https://github.com/docToolchain/docToolchain/issues/259#issuecomment-469430444) @rdmueller mentioned the `bin/doctoolchain` idea
[] we had the git submodule approach
[] clone the repo
[] ...

**Comment by jakubjab** [comment]
I think to make a decision we could list what alternatives we have, and rate them against the quality goals we have: Confidentiality, Ease of Use, Maintainability.

Regarding alternatives, I see the following:

A) We could focus on delivering docToolchain as a standalone tool and remove other options. Maintainability is better since there are fewer options to test.
Ease of Use is better since:
- users are not confused with multiple options how to use the tool
- a standalone tool can be integrated with any toolchain, either using gradle or other systems (including other versions of gradle)
- a standalone tool can be upgraded independently of the main toolchain (less coupling)

B) We could focus on delivering docToolchain as a gradle addon and remove other options. Maintainability is better since there are fewer options to test.
Ease of Use is better since:
- users are not confused with multiple options how to use the tool
- they use the same gradle as they do for main development
Ease of Use is worse since:
- a gradle addon can be only used if the main toolchain uses the same gradle

Whichever way, I think the initialization should be decoupled from the main tool. Downloading arc42 template happens maybe once per project, while the building happens hopefully after each commit, many times per day. I don't see a good reason to keep both pieces of code together, making the maintenance more complex. We could have a separate repository with initialization script, or just describe it in the manual. In the overall cost of the project, the cost of downloading a template is negligible.

In other words, I think considering the limited resources we should cut the scope and focus on really key features, bringing most value.

Personally, for me the value is in having a standalone tool (option A) that can be run in CI frequently and is stable, tested and bulletproof. I want to focus on the contents, a little on the configuration, not on the tool. I don't care which language it's written in.

I'm open for other opinions.

**Comment by rdmueller** [comment]
ok. I will try to come up with a draft of a design decision / ADR

**Comment by jakubjab** [comment]
Then some more thoughts: maybe it would be a good idea to split different tasks more clearly into separate tools and define clear interfaces between them. For example to define:

- `bin/dt-asciidoc-to-pdf` converts from asciidoc to PDF
- `bin/dt-asciidoc-to-html` converts from asciidoc to HTML
- `bin/dt-html-to-confluence` exports from HTML to Confluence
- `bin/dt-html-to-confluence` exports from HTML to Confluence
- `bin/dt-asciidoc-to-docbook` converts from asciidoc to DocBook
- `bin/dt-docbook-to-epub` converts from DocBook to epub

and so on.

By separating I mean that each tool has a separate directory, can be deployed or skipped if not needed, can be tested separately, can be even rewritten in a different language if we find that more suitable (e.g. plain bash), can require separate dependencies, can evolve independently.

By clear interfaces I mean the input/output format (usually directory structure and files, or some state of some external system like Confluence) and configuration options. Input/output format can be documented by example, by having several examples in the test. Configuration options should be documented in the tool itself as command line options or configuration files.

**Comment by rdmueller** [comment]
@jakubjab thanx for your thoughts. I would like to add some requirements:

- when cloned, you should be able to fully build the project without settings up any extra dependencies.
This means we need at least a small wrapper which downloads docToolchain
- Task dependencies. I would like to make the docs depend on the test. This way, we can include test results in the docs and ansure that they work
- we also have to consider the operating system. Some tasks are windows only and thus hard to run in a docker environment.
- when we speak of docker: the build has to run in virtual environments where docker is not available. So if we take a docker based solution (which is  nice) into consideration, we have to also create a fallback

Regarding maintenance and testing: my dream is to set up some tutorials which can be fully automated tested, but I still have no clue what would be the best way to go.

Regarding the initialisation: it would be great if we could come up with some kind of "convention over configuration" approach and a really usable tool. It should be as simple as possible to use docToolchain.
I've already started to add some help messages which replace some error messages to help people get started. Currently, many of the users are power users (and I am glad about it!). They read the code to see what is going wrong. In the future, we can't expect this anymore.

**Comment by jakubjab** [comment]
Just to clarify:

> when cloned, you should be able to fully build the project without settings up any extra dependencies

Cloned what? docToolchain or a project using asciidoc for documentation?

**Comment by gernotstarke** [comment]
our role-model could be [pandoc](https://pandoc.org/):

[] it's easy to install - and nobody needs to know about its (magical) inner workings
[] you can configure output options with command-line flags (such flags could imho replace the various tools mentioned by @jakubjab above)

It's a library, with the command-line tool as a lightweight wrapper.

In contrast to pandoc, we currently have an inevitable dependency on a Java Runtime Environment (but that's ok to keep from my perspective).

The dependency on gradle, on the other hand, might limit acceptance by teams using maven, or working with bundler, rake or other build-tools.

**Comment by jakubjab** [comment]
I agree pandoc is a good example, it’s a simple tool that is doing its single job of converting between formats very well. If the command line interface is based on options or separate binaries is to be decided, I don’t mind either way. Dependencies of the tool are acceptable as long as they are managed automatically and don’t interfere with the environment too much.

**Comment by jakubjab** [comment]
Using pandoc as example, I would like to use docToolchain in the similar way:

`doctoolchain -f asciidoc -t html mydoc.adoc -o mydoc.html`

Then if I wanted to include Confluence:

`doctoolchain -f html -t confluence mydoc.html -confluenceapi=... -confluenceparent=...`

No configuration files, no paths hidden in the tool.

**Comment by jakubjab** [comment]
@rdmueller - any comments? I would love to start working on the new interface for the standalone tool (even if I don't know where to start exactly), but first we need a clear decision for the overall direction.

**Comment by rdmueller** [comment]
sorry, I am currently at JavaLand and still need some more time...

**Comment by blackfoxoh** [comment]
Hi,
this discussion is interesting even I'm not sure if understood everything in detail. Maybe I can contribute a little bit with what I experienced with doctoolchain so far:
- I heard of doctoolchain at jax and from the articles in javamagazin and found it very interesting
- My intentions for doctoolchain: built up arc42 docu from scratch but also use it for other documentation like  usersguide etc. which should be maintained in sync with the source
- I downloaded a zip-copy of the gitrepo because we don't use git so far
- in the next steps I struggeld a lot in getting everythin up and running:
my big problem is that we are behind a corporate firewall and all the magic doctoolchain offers with initialization failed. I have little experience in gradle but actually not very much. We use a local artifactory repo as a proxy for librarys. With lot's of trying I found several places to edit so all the gradle stuff and libraries could be downloaded. Now I asked myself: when doctoolchain evolves further I will want to keep up with this but when replacing the doctoolchain with it's new version I have to reconfigure all theese things again.
For the moment I stopped my test of doctoolchain, because of this issues and time-reasons as well. I plan to proceed somewhen. I think doctoolchain is a great idea and helps already a lot in configuring a complex toolchain - none the less there might be an initial barrier in getting startet behind a firewall which prevents lot of users from getting really started. A central place for configuration (if feasible) or at least hints where to modify would help a lot :)

**Comment by rdmueller** [comment]
@blackfoxoh thanx for this feedback! It is very valuable.
Maybe we should add some tests for network connectivity etc to make it easier to get started behind a firewall. I've had the same problems in the past and have written down my experience in two blog posts:

https://docs-as-co.de/news/enterprise-edition/
https://docs-as-co.de/news/enterprise-edition2/

Maybe this helps a bit...

**Comment by Chessray** [comment]
Hi,

As someone just starting their (practical) arc42 journey, I'm very interested in the toolchain as well. We're introducing arc42 to our current client and would like to use doctoolchain in our CI environment. Now I have to point out that this particular client is actually a full .NET house and we're running in Azure DevOps. While we could use the Gradle tooling in principle (and I feel confident enough to get this up and running), the client would probably not be very happy in maintaining the setup. For now, I only cloned the doctoolchain GitHub repo onto my box and I'm using it to download the adoc template into the various code repos. But we started this work only over the last few days and haven't really looked yet into how we will included the output generation into a possible CI build.

Has there been any movement in this area since March and how might we be able to contribute?

**Comment by Guchman** [comment]
Hi, I'm just trying to use doctool and it's not goes smooth.
I think there is two main usecases:
1) use doctoolchain as separate cli tool
2) use doctoolchain as task in build tool like gradle

First case need binary
Second one needs some maven/gradle plugin to run that binary/jar.

Now it's seems like mix of both. I need to clone repo, and then use gradle from it to run tasks.
Is there any good reason to use gradle tasks to RUN doctoolchain tasks? 
I think use gradle as your program runner should be considered as anti-pattern.
Any way, I prefer to use plugin:) but it's would be wonderful separate plugin and tool itself, so I think maven folks want to use it as well.

**Comment by jakubjab** [comment]
Yes, it is a mix because of historical reasons. I'm also using doctoolchain as a separate CLI tool in a cmake build system, it should be usable from any build system that can execute commands. I don't like the dependency on gradle, but it's not a big problem either. I just treat that as a quite big script, which works fine if I provide dependencies. You can also try https://github.com/docToolchain/docker-image if you prefer.

**Comment by Guchman** [comment]
But still, if I want to use docToolChain as task in my project, how to achieve this?

**Comment by rdmueller** [comment]
Hi everyone...
I've now seen docToolchain used by several teams with different backgrounds in different ways and I think I have come to a conclusion on how we should evolve docToolchain from here.

Let's first describe the vision.
Imagine:
- docToolchain is just a cli tool (dtc) with a small wrapper (dtcw) which is the only thing you have to copy to your repository and configure it
- only dependency (like gradlew or mvnw) is java on your machine. Maybe we can even get rid of this.
- the wrapper is configured for a specific version. If this version is not found in the user`s home directory, it will be downloaded from a standard or configured location. No pollution in the project
- this way, if anybody clones the repository, all dependency will be available, no additional manual install
- the project repository itself only contains the wrapper and documents (and additional project code). No `gradle.build`, no docToolchain sources
- if you need to change the behaviour of the docToolchain scripts or maybe the pdfStyle, the will a command like `dtcw copyScripts` or `dtcw copyPdfStyle` which copies these artefacts from the docToolchain project to your repository (folder `/docToolchain`) for you to overwrite. Delete files which you don't want to change to fall back to the standard file

As a result,
- your project will not be polluted anymore
- cli works - it will also work in a container environment
- no git submodule anymore (Yeah!)
- each project still knows which dtc version is needed
- it will be easy to use

The solution will build on the cli install created by @jakubjab but use some special mechanismns:
- the wrapper will automaticall download docToolchain as `.zip`
- docToolchain will be installed to `$USER_HOME/.docToolchain/dtc-version.zip`
- the wrapper will unzip it to `$projectDir/build/docToolchain/.` it will be executed from there.
- the wrapper will also copy (end hence overwrite some files) `$projectDir/docToolchain/.` to `$projectDir/build/docToolchain/.` . This will guarantee maximum flexibility
- if the wrapper is unable to find a java version, it will output a help text on how to install java via sdkman

What do you think?

**Comment by jakubjab** [comment]
I like the general idea.

Some comments:

> the project repository itself only contains the wrapper and documents

I would prefer the wrapper outside of the project repository if possible, but it's still better than having whole doctoolchain included.

> if you need to change the behaviour of the docToolchain scripts or maybe the pdfStyle...

I would split it.
If you need to change the pdfStyle, fonts, colors, etc., all this should be possible to specify in the command line, either directly, or by giving path to some configuration file (only for cases unsuitable for the command line, e.g. fonts).
If you need to change the behaviour of the docToolchain scripts, then the first step should be creating a fork on GitHub and changing them, preferably in non-breaking way. If that's not possible, then `dtcw copyScript` could be an option, but I still think you can have your patches to the code and use the git power to apply them after upgrading to a new version (with upgrades to patches as needed).

> the wrapper will automaticall download docToolchain as .zip

I would strongly argue for any existing package manager, see https://en.wikipedia.org/wiki/List_of_software_package_management_systems and choose one. I have good experience with `pkgsrc`, would be really good to just write `pkgin install doctoolchain` and have the wrapper installed. I'm sure for Windows there are some similar solutions. In our company `pkgsrc` is used to distribute also our internal software, it's fully configurable.

> the wrapper will unzip it to `$projectDir/build/docToolchain/`

This I don't understand, I thought the project directory was supposed to be clean?

In general, the strongest constraint is that docToolchain is used in the automatic build tool chain, and the whole project directory is read only (to prevent accidental/hostile modification of code when building it, to reduce security threats). All the intermediate and output files need to be somewhere else. The other constraint is that on the same machine, with the same user, multiple builds are running. Each of them needs to be independent, so either we use standards like generating random temporary file names, or require each build to have a unique identifier which can be used.

If those constraints are too strong at the start, we can make some limitations and get there iteratively.

**Comment by rdmueller** [comment]
@jakubjab thanx for the valuable input

> I would prefer the wrapper outside of the project repository if possible...

I think I do understand. I guess I will set up the wrapper as additional repository and current functionality will be kept. This will result in a wrapper which could make it easier to use dtc for some people, but keep the functionality that you can install dtc simply as command line tool without the wrapper.

> I would strongly argue for any existing package manager...

currently I see no benefit of an existing package manager. In contrast, I see the need for people wihtouth an isntalled package manager to easily use dtc. So I am not sure about it.

>>the wrapper will unzip it to $projectDir/build/docToolchain/
>This I don't understand, I thought the project directory was supposed to be clean?

Interesting thoughts which make sense. To be honest, `/build/` should be configurable. But that's what most java repositories do: store all temporarily created artefacts in the `build` folder.

**Comment by jakubjab** [comment]
> I see the need for people wihtouth an isntalled package manager to easily use dtc. So I am not sure about it.

We could have both - if we have raw downloading wrapper, I guess it should be easy to wrap it in some package managers as a next step. See for example https://github.com/Homebrew/homebrew-core/blob/master/Formula/asciidoc.rb or https://github.com/Homebrew/homebrew-core/blob/master/Formula/asciidoctor.rb for Homebrew formulas.

> store all temporarily created artefacts in the build folder

As long as it's configurable it's fine for me.

**Comment by rdmueller** [comment]
Great News! : docToolchain 1.2.0-beta has just been released on sdkman :-)

@jakubjab thanx for your work to turn docToolchain into a CLI tool. This helped a lot to do this first release

The first tests are quite positive. It works. It would be great if you could also run some tests to see if everything works as expected.

**Comment by mh182** [comment]
Closed since it is already implemented.
